err#8038: Index.shift() gives confusing error message when no Datetim…

hamedhsn opened the pull request:
…eIndex

closes #8038

maximilianr commented:
Is this the behavior you want though, vs. a clearer error message? This doesn't seem like it's shifting the index; it seems like it's adding 1...
```python
        result = pd.Index(range(0,10,2)).shift(1)
        expected = pd.Index([1, 3, 5, 7, 9], dtype='int64')
```

jreback commented:
@hamedhsn I don't this should even work on a non-datetimelike index as it doesn't make any sense. So we should prob raise a ``TypeError``.

jreback commented on pandas/core/index.py:
```diff
@@ -1411,7 +1411,11 @@ def shift(self, periods=1, freq=None):
-        offset = periods * freq
+        try:
+            offset = periods * freq
+        except TypeError:
```
no, pls. raise a ``NotImplementedError`` for all ops in ``core/index.py`` for ``.shift``. Then define this in ``tseries/base/DatetimeIndexOpsMixIn``

jreback commented:
pls add a test and a whatsnew note in 0.17.1

hamedhsn commented:
@jreback 
1)why do I need to change tseries/base/DatetimeIndexOpsMixIn ? 
2)where should I add tests?

jreback commented:
I indicated above. The ``.shfit` in ``core/index.py`` applies to ``Index/Int64Index`` etc. These are not valid for shifting and so should raise ``NotImplementedError``. Then make the actual change in ``tseries/base/DatetimeIndexOpsMixIn`` which is the base class for the datetimelikes (``DatetimeIndex,TimeDeltaIndex,PeriodIndex``) where shifting makes sense

hamedhsn commented:
sure, but there is no issue with DatetimeIndex using shift() with or without freq. that was my question.

the error happens for Index/Int64Index which i changed it to raise NotImplementedError. 
btw where do i have to add the related tests?

jreback commented:
@hamedhsn you can add tests in ``test_index`` (though the timeseries ones are prob in ``tseries/test_base``)

Commit:
err#8038 :add fix, note and test

err#8038 : add fix, note and test

err#8038 : add fix, note and test

fix added with tests

modification

Index.shift() gives confusing error message when no DatetimeIndex

test, fix note

change tests

add issue number

test change

test change

add tests

add test

hamedhsn commented:
@jreback this also should be fine

maximilianr commented:
@jreback @hamedhsn I think this behavior is confusing. I think there are some reasons to shift a non-DateTime-like index, but it shouldn't be done with a `freq`.

Do you guys think this is clear and expected behavior?

```python
        idx = Index(range(5))   
        result = idx.shift(periods=2, freq=2)
        expected = pd.Index(range(4, 9))
        self.assert_index_equal(result, expected)
```

A version I could see being helpful is to shift by a number of items. i.e. this:
```python
        idx = Index(range(5,10))   
        result = idx.shift(1)
        expected = pd.Index(range(6, 11))
        self.assert_index_equal(result, expected)
```
...although I'm more against the current implementation than I am fighting for this specific implementation.

As ever, open to being wrong though.

jreback commented:
@maximilianr that would be a major change to add a freq to a non-timeseries Index. So wouldn't want to support that (here).

maximilianr commented:
@jreback That's exactly what I'm saying. This PR enables `freq` for shifting non-timeseries index (or am I woefully mistaken)? This is a copy paste from the new test code:
```python
        idx = Index(range(5))   
        result = idx.shift(periods=2, freq=2)
        expected = pd.Index(range(4, 9))
        self.assert_index_equal(result, expected)
```

jreback commented:
that's out of scope for this PR
would be unexpected as freq is not a property of an Index 

you are free to propose it but make an issue

maximilianr commented:
I don't think I'm being clear. Your comment [here](https://github.com/pydata/pandas/pull/11211#discussion_r41131777) would fix the concern, but that hasn't been added. 
Or I'm likely missing something, so I'll leave it for now

jreback commented on doc/source/whatsnew/v0.17.1.txt:
```diff
@@ -26,7 +26,7 @@ Enhancements
 Other Enhancements
 ^^^^^^^^^^^^^^^^^^
-
+- raising ``NotImplementedError`` for ``Index.shift()`` when no ``DatetimeIndex`` is used.
```
say: raise ``NotImplementedError`` in ``Index.shift`` for non-supported index types. (and include the issue number)

jreback commented on pandas/core/index.py:
```diff
@@ -1416,7 +1416,11 @@ def shift(self, periods=1, freq=None):
+            offset = periods * freq
+        except TypeError:
+            raise NotImplementedError("freq should be defined for %s" % self.dtype)
+
```
simply raise ``NotImplementedError`` here always; define this function instead in ``tseries/base/DatetimeIndexOpsMixin`` (w/o the ``NotImplemenetdError``)

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2852,6 +2852,18 @@ def test_constructor_corner(self):
         with tm.assertRaisesRegexp(TypeError, 'casting'):
             Int64Index(arr_with_floats)
 
+    def test_notimplementederror_shift_int64(self):
```
move the ``NotImplementedErrorr`` case to the ``Base`` class so its tested on all index types; while in the datetimelike override

hamedhsn commented:
@jreback is this implementation correct? let me know if I need to anything

jreback commented on doc/source/whatsnew/v0.17.1.txt:
```diff
@@ -27,10 +27,14 @@ Enhancements
 Other Enhancements
 ^^^^^^^^^^^^^^^^^^
+<<<<<<< HEAD
+- raising ``NotImplementedError`` for ``Index.shift()`` when no ``DatetimeIndex`` is used.
```
rebase artifcacts

jreback replied:
add the issue number

jreback replied:
put in API changes as this 'worked' before

jreback commented on pandas/core/index.py:
```diff
@@ -1412,12 +1412,7 @@ def shift(self, periods=1, freq=None):
-
-        offset = periods * freq
-        return Index([idx + offset for idx in self], name=self.name)
+        raise NotImplementedError
```
raise with a helpful message (e.g. say something about not supported for this type (e.g. show the type))

jreback commented on pandas/tseries/base.py:
```diff
@@ -505,9 +505,10 @@ def shift(self, n, freq=None):
                 freq = frequencies.to_offset(freq)
-            result = Index.shift(self, n, freq)
+            offset = n * freq
+            result = Index([idx + offset for idx in self], name=self.name)
```
``result = self + offset`` is MUCH faster

jreback commented on pandas/tseries/tests/test_base.py:
```diff
@@ -45,6 +45,22 @@ def test_ops_properties_basic(self):
         self.assertRaises(AttributeError, lambda : s.weekday)
 
+    def test_shift(self):
+        idx = Index(range(5))
```
add the issue number as a comment

jreback commented:
a lot closer....still a couple of comments

hamedhsn commented:
@jreback applied your comments

jreback commented on pandas/core/index.py:
```diff
@@ -1413,12 +1413,7 @@ def shift(self, periods=1, freq=None):
-
-        offset = periods * freq
-        return Index([idx + offset for idx in self], name=self.name)
+        raise NotImplementedError("Not supported for type %s" % self.dtype)
```
use ``type(self).__name__``

jreback commented:
one more comment. pls squash as well.

hamedhsn commented:
@jreback did that. let me know if I need to do anything else. Thanks

jreback commented on pandas/tseries/tests/test_base.py:
```diff
@@ -45,6 +45,24 @@ def test_ops_properties_basic(self):
         self.assertEqual(s.day,10)
         self.assertRaises(AttributeError, lambda : s.weekday)
 
+    def test_shift(self):
```
move this to ``tests/test_index`` (I make have told you to put in in ``tseries/test_base`` before, sorry), where the ``NotImplementedError`` is raised (this should go in the ``Base`` class ), then in ``Datetimelike`` you override to have it tested in the 3 datetimelikes

hamedhsn replied:
@jreback hope I got it right..I moved the test for base(NotImplementedError) to index.

jreback replied:
move these to ``tests/test_index.py``

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2815,6 +2815,13 @@ def testit():
         assertRaisesRegexp(ValueError, "^Length", testit)
 
+    def test_shift_index(self):
+        idx = Index(range(5))
```
add the issue number as a commet

jreback replied:
this should be in ``Base`` and use ``self.create_index()`` (override for the ``DatetimelikeIndexes`` to actually do a shift)

jreback commented:
comments; needs a rebase

hamedhsn commented:
@jreback applied the comments

jreback commented on pandas/tseries/tests/test_base.py:
```diff
@@ -45,6 +45,14 @@ def test_ops_properties_basic(self):
         self.assertEqual(s.day,10)
         self.assertRaises(AttributeError, lambda : s.weekday)
 
+    def test_shift_index(self):
```
move this to ``test_index.py`` inside that ``Base`` class

further you will need to create the index with ``self.create_index()``

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2815,6 +2815,19 @@ def testit():
 
+    def test_shift(self):
+        # test shift for datetimeIndex and non datetimeIndex
+        # err8083
```
create the index with ``self.create_index()`` and put in the ``DateTimelike``

hamedhsn commented:
@jreback should be fine now hopefully

jreback commented:
@hamedhsn if you would like to update according to comments would be geat

hamedhsn commented:
@jreback I think I did, used self.create_index(), put that in DateTimelike. and for the other one moved that to based class in test_index.py and used self.create_index(). and rebased that.

jreback commented on pandas/tests/test_index.py:
```diff
@@ -3324,6 +3332,22 @@ def setUp(self):
+        # test shift for datetimeIndex and non datetimeIndex
+        # err8083
+
+        drange = self.create_index()
```
this is ONLY true for DatetimeIndex. Need separate expectation for TimedeltaIndex/PeriodIndex (IOW separate methods).

hamedhsn commented:
@jreback added the other two tests under their own test classes

jreback commented:
xref #11452

jreback commented on pandas/tests/test_index.py:
```diff
@@ -53,6 +53,14 @@ def test_pickle_compat_construction(self):
+        idx = self.create_index()
+        self.assertRaises(NotImplementedError, idx.shift, 1)
+
+        result = idx.shift(periods=2, freq=2)
```
take out this call right here as it will fail by definition (leave the ``assertRaises``)

jreback commented on pandas/tests/test_index.py:
```diff
@@ -3425,6 +3433,22 @@ def setUp(self):
+        self.assert_index_equal(result, expected)
+
+        result = drange.shift(3, freq='2D')
+        expected = DatetimeIndex(['2013-01-07', '2013-01-08', '2013-01-09', '2013-01-10',
```
can you add negative shift & a 0 shift test as well (which should return identically)

jreback commented:
@hamedhsn ok, tests look good (you just need to remove an error line and they would pass). pls add the other requested tests and good 2 go.

hamedhsn commented:
@jreback applied your comments.rebased it

jreback commented:
merged via https://github.com/pydata/pandas/commit/a05eceba6c869217c5c6b8bf2bb57ff5cef8e681

thanks!

danger89 commented:
It seems I still have this issue see my comment in: #11452. I'm using pandas v0.19.2 via pip.
